### PR TITLE
feat: add pause and resume jetstream consumer

### LINKED ISSUE #1569: Implement Consumers Pause
### Proposed change

Implement the new API for pausing consumers.

The Server PR with related change has been merged here: https://github.com/nats-io/nats-server/pull/5066
Schemas PR, which is probably the best source of information how to implement the feature:
https://github.com/nats-io/jsm.go/pull/522/files
ADR issue: https://github.com/nats-io/nats-architecture-and-design/issues/266

## Details
**Request and response schemas:**
```go
// io.nats.jetstream.api.v1.consumer_pause_request
type JSApiConsumerPauseRequest struct {
	PauseUntil time.Time `json:"pause_until,omitempty"`
}

// io.nats.jetstream.api.v1.consumer_pause_response
type JSApiConsumerPauseResponse struct {
	JSApiResponse
	Paused         bool          `json:"paused"`
	PauseUntil     time.Time     `json:"pause_until"`
	PauseRemaining time.Duration `json:"pause_remaining,omitempty"`
}
```
Consumer info includes 2 new fields:

```go
Paused bool json:"paused,omitempty"
PauseRemaining time.Duration json:"pause_remaining,omitempty"
```
The paused state and time time would need to be persisted to the raft layer such that server restarts would not unpause paused consumers. This is done using the consumer configuration that has a new value:
```go
PauseUntil time.Time json:"pause_until,omitempty"
```
If in doubt, refer to the Server/Schema PR

Feature be available in the server version 2.11
### Hints
Probably the easiest way to implement it, is to look how calls to methods for creating / deleting streams and consumers are done.

### Use case

It is difficult to schedule maintenance on central resources on a large distributed system where 100s or 1000s of clients are accessing data in a stream.

This change allows to pause a Consumer such that it appears healthy but just doesnt deliver any messages.

During the pause maintenance can happen and resources accessed by clients will not be under constant pressure, later the stream can be unpaused and work will continue.

This happen without impacting running clients - other than they would see pending messages in stream info but not get any deliveries.

Applies to push and pull consumers.

### Contribution

_No response_

### PR DESCRIPTION
**Pull request by yordis**
closes #1569

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -3,13 +3,20 @@ module github.com/nats-io/nats.go
 go 1.20
 
 require (
-	github.com/klauspost/compress v1.17.2
+	github.com/golang/protobuf v1.5.3
+	github.com/klauspost/compress v1.17.6
+	github.com/nats-io/nats-server/v2 v2.10.11
```
Comment (yordis) [inline go.mod:8]: I am not sure why, but this is required in order to compile, I ran `go mod tidy`

Comment (wallyqs) [reply go.mod:8]: you need to run with `go test -modfile=go_test.mod`

File: go_test.mod [hunk 2]
```diff
@@ -8,15 +8,20 @@ require (
 	github.com/nats-io/nats-server/v2 v2.10.11
 	github.com/nats-io/nkeys v0.4.7
 	github.com/nats-io/nuid v1.0.1
+	github.com/stretchr/testify v1.8.0
```
Comment (yordis) [inline go_test.mod:11]: 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺

File: jetstream/consumer.go [hunk 3]
```diff
@@ -317,6 +318,42 @@ func deleteConsumer(ctx context.Context, js *jetStream, stream, consumer string)
...
+	}
+	if err := validateConsumerName(consumer); err != nil {
+		return err
+	}
+	subject := apiSubj(js.apiPrefix, fmt.Sprintf(apiConsumerPauseT, stream, consumer))
+
+	var resp consumerPauseResponse
+	req, err := json.Marshal(consumerPauseRequest{
+		PauseUntil: pauseUntil,
+	})
```
Comment (yordis) [inline jetstream/consumer.go:334]:
```suggestion
	req, err := json.Marshal(consumerPauseRequest{
		PauseUntil: nil,
	})
```

Should this work with `nil` PauseUntil value?

It didn't work for me; I had to change to a `time.Now()` value instead.

Should I send a empty body when I want to `resume` right away?

Need help here!

**Comment by yordis** [comment]
I need help with:

1. What should be the final public API (should I remove `resume` and expect people to call `Pause` with the `time.Now()`)
2. Verify that the payload is correct by not allowing `nil` time

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,6 +1,8 @@
 # Contributing
 
-Thanks for your interest in contributing! This document contains `nats-io/nats.go` specific contributing details. If you are a first-time contributor, please refer to the general [NATS Contributor Guide](https://nats.io/contributing/) to get a comprehensive overview of contributing to the NATS project.
+Thanks for your interest in contributing! This document contains `nats-io/nats.go` specific contributing details. If you
```
Comment (wallyqs) [inline CONTRIBUTING.md:3]: fixes to the contributing doc could be a separate commit in a different by itself, might be good to take it out of this PR then rebase

**Comment by ripienaar** [comment]
In jsm.go - that is the underlying library for the CLI I added:

 * `PauseUntil(deadline time.Time)` as a create option to create the consumer paused
 * `Pause(deadline time.Time)` to pause a already created consumer
 * `Resume()` to resume a paused consumer
 * `IsPaused()` on a consumer - which I now notice I forgot to add

CLI got:

```
  consumer pause    Pause a consumer until a later time
  consumer resume   Resume a paused consumer
```

So seems good to me to have pause and resume behaviors in APIs, the original ask from the server was to have a Resume API also but we decided to take a bit of a shortcut with the nil time behavior.  I think client facing be good to have Pause/Resume both.

File: jetstream/jetstream.go [hunk 5]
```diff
@@ -183,6 +183,12 @@ type (
 		// DeleteConsumer removes a consumer with given name from a stream.
 		// If consumer does not exist, ErrConsumerNotFound is returned.
 		DeleteConsumer(ctx context.Context, stream string, consumer string) error
+
+		// PauseConsumer pauses a stream. If pauseUntil is provided, the stream will be paused until that time.
```
Comment (piotrpio) [inline jetstream/jetstream.go:187]: If you want to pause a consumer, you have to provide `pauseUntil`, so the comment is a bit misleading. Also, I would change the signature to accept value, not a pointer.

Also, PauseConsumer pauses a **consumer**, not a stream

File: jetstream/jetstream.go [hunk 6]
```diff
@@ -183,6 +183,12 @@ type (
...
+		PauseConsumer(ctx context.Context, stream string, consumer string, pauseUntil *time.Time) error
+
+		// ResumeConsumer resumes a paused stream.
```
Comment (piotrpio) [inline jetstream/jetstream.go:190]:
```suggestion
		// ResumeConsumer resumes a paused consumer.
```

File: jetstream/stream.go [hunk 7]
```diff
@@ -101,6 +101,12 @@ type (
 		// If consumer does not exist, ErrConsumerNotFound is returned.
 		DeleteConsumer(ctx context.Context, consumer string) error
 
+		// PauseConsumer pauses a consumer.
+		PauseConsumer(ctx context.Context, consumer string, pauseUntil *time.Time) error
```
Comment (piotrpio) [inline jetstream/stream.go:105]: Same as above, I would not use pointer here.

File: go_test.mod (hunk 2 above)
Comment (yordis) [reply go_test.mod:11]: 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺

So much work without it for such minor thing

🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺 🥺

**Comment by yordis** [comment]
Hey guys, any update over here?

**Review by piotrpio — APPROVED** [review]
Great work, LGTM!

I changed the target branch to `v2.11-dev` since we don't want to merge to main until the feature is released on the server.

**Comment by yordis** [comment]
Waiting to be merge so I move to the Elixir version
